### DISCUSSION TITLE
How to test for chained exceptions?

### DISCUSSION THREAD

**Discussion by GlenNicholls** [pytest-dev/pytest#9979] [Q&A] [answered]
I wrote a JSON schema that defines the expected data structure for a configuration file and I have an exception that gets raised by a third-party package (jsonschema `ValidationError`). The exception value contains info about what attributes failed validation. In my tests, I would like to make sure my schema is properly catching all issues possible. But, I need to tell the user with a new exception which file has the error, and thus, I'm catching this exception and raising a new one that points to the JSON file causing the error. In essence, the following:

``` python
def parse(file):
  data = json.from_file(file)
  try:
    validate(data)
  except ValidationError as exc:
    raise NewError(f"File failed validation '{<some_file>}'") from exc
  else:
    return data
```

The problem I'm running into is how do I test that I'm correctly handling each error condition? I would like to make sure that I am matching against the correct `ValidationError` message while running my tests to ensure that my JSON schema is actually defining the attributes of the JSON data correctly. Basically, the below is useless in pytest other than understanding the exception was raised, however, it doesn't test that my schema defined some properties correctly:

``` python
with pytest.raises(NewError, match="File failed validation")
  parse(<some_file>)
```

What I can do is add it to my new error like so:

``` python
try:
  validate(data)
except ValidationError as exc:
  raise NewError(f"{exc}\nFile failed validation '{<some_file>}'") 
```

But this doesn't really solve my issue as critical information may be lost by keeping the value of the third-party error and not the metadata of the traceback. 

In short, what I'm asking is how do I check the chained traceback for a specific string/pattern? `pytest.raises` has the `match` keyword, but it only works on the caught exceptions message, not other messages in the traceback.

**Comment by portyu9** [comment, marked as answer]
Hi @GlenNicholls 

Because you are using explicit exception chaining:

```python
raise NewError(...) from ValidationError(...)
```

the original exception is available as `__cause__`.

You can assert both exceptions without duplicating the inner message into `NewError`:

```python
with pytest.raises(NewError, match="File failed validation") as exc_info:
    parse(file)

cause = exc_info.value.__cause__

assert isinstance(cause, ValidationError)
assert "expected validation message" in str(cause)
```

`__cause__` is specifically the exception supplied with `raise ... from ...`.

For an exception raised while another exception is merely being handled, without `from`, Python instead records it in `__context__`.

Please mark this answer as accepted if it helped, thank you.
